rtl/eclrtl/rtldynfield.cpp

```diff
@@ -510,3 +510,481 @@ extern ECLRTL_API int getFieldNum(const char *fieldName, IOutputMetaData &  meta
         return r->getFieldNum(fieldName);
     return -1;
 }
+enum FieldMatchType {
+    // These are in order of pain
+    // On a field, exactly one of the above is set, but translator returns a bitmap indicating
+    // which were required (and we can restrict translation to allow some types but not others)
+    match_perfect     = 0x00,    // exact type match - use memcpy
+    match_link        = 0x01,    // copy a nested dataset by linking
+    match_truncate    = 0x02,    // dest is truncated copy of source - use memcpy
+    match_extend      = 0x04,    // dest is padded version of source - use memcpy and memset
+    match_typecast    = 0x08,    // type has changed - cast required
+    match_none        = 0x10,    // No matching field in source - use null value
+    match_recurse     = 0x20,    // Use recursive translator for child records/datasets
+    match_fail        = 0x40,    // no translation possible
+};
+
+inline constexpr FieldMatchType operator|(FieldMatchType a, FieldMatchType b) { return (FieldMatchType)((int)a | (int)b); }
+inline FieldMatchType &operator|=(FieldMatchType &a, FieldMatchType b) { return (FieldMatchType &) ((int &)a |= (int)b); }
+
+class GeneralRecordTranslator
+{
+public:
+    GeneralRecordTranslator(const RtlRecord &_destRecInfo, const RtlRecord &_srcRecInfo)
+    : destRecInfo(_destRecInfo), sourceRecInfo(_srcRecInfo)
+    {
+        matchInfo = new MatchInfo[destRecInfo.getNumFields()];
+        createMatchInfo();
+    }
+    ~GeneralRecordTranslator()
+    {
+        delete [] matchInfo;
+    }
+    void describe(unsigned indent = 0) const
+    {
+        for (int idx = 0; idx <  destRecInfo.getNumFields(); idx++)
+        {
+            const RtlFieldInfo *field = destRecInfo.queryField(idx);
+            const MatchInfo &match = matchInfo[idx];
+            if (match.matchType == match_none)
+                DBGLOG("%*sNo match for field %s", indent, "", field->name);
+            else
+            {
+                DBGLOG("%*sMatch (%d) for field %d for field %s", indent, "", match.matchType, match.matchIdx, field->name);
+                if (match.subTrans)
+                    match.subTrans->describe(indent+2);
+            }
+        }
+        if (!canTranslate())
+            DBGLOG("Translation is NOT possible");
+        else if (needsTranslate())
+            DBGLOG("Translation is possible");
+        else
+            DBGLOG("Translation is not necessary");
+    }
+    size32_t translate(ARowBuilder &builder, size32_t offset, const byte *sourceRec) const
+    {
+        unsigned numOffsets = sourceRecInfo.getNumVarFields() + 1;
+        size_t variableOffsets[numOffsets];
+        RtlRow sourceRow(sourceRecInfo, sourceRec, numOffsets, variableOffsets);
+        for (int idx = 0; idx < destRecInfo.getNumFields(); idx++)
```

Comment:
If you can't do it completely upfront, resizing to the minimum size required (and doing the same on all subsequent calls, which is slightly tricky) means you only need to resize for variable width targets.